src/zenml/integrations/kubernetes/pod_settings.py

```diff
@@ -0,0 +1,289 @@
+
+import re
+from datetime import date, datetime
+from typing import TYPE_CHECKING, Any, Dict, List, Type, Union, cast
+
+from pydantic import validator
+
+from zenml.config.base_settings import BaseSettings
+
+if TYPE_CHECKING:
+    from kubernetes.client.models import V1Affinity, V1Toleration
+
+
+class KubernetesPodSettings(BaseSettings):
+    """Kubernetes Pod settings.
+
+    Attributes:
+        node_selectors: Node selectors to apply to the pod.
+        affinity: Affinity to apply to the pod.
+        tolerations: Tolerations to apply to the pod.
+    """
+
+    node_selectors: Dict[str, str] = {}
+    affinity: Dict[str, Any] = {}
+    tolerations: List[Dict[str, Any]] = []
+
+    @validator("affinity", pre=True)
+    def _convert_affinity(
+        cls, value: Union[Dict[str, Any], "V1Affinity"]
+    ) -> Dict[str, Any]:
+        """Converts Kubernetes affinity to a dict.
+
+        Args:
+            value: The affinity value.
+
+        Returns:
+            The converted value.
+        """
+        from kubernetes.client.models import V1Affinity
+
+        if isinstance(value, V1Affinity):
+            return serialize_kubernetes_model(value)
+        else:
+            return value
+
+    @validator("tolerations", pre=True)
+    def _convert_tolerations(
+        cls, value: List[Union[Dict[str, Any], "V1Toleration"]]
+    ) -> List[Dict[str, Any]]:
+        """Converts Kubernetes tolerations to dicts.
+
+        Args:
+            value: The tolerations list.
+
+        Returns:
+            The converted tolerations.
+        """
+        from kubernetes.client.models import V1Toleration
+
+        result = []
+        for element in value:
+            if isinstance(element, V1Toleration):
+                result.append(serialize_kubernetes_model(element))
+            else:
+                result.append(element)
+
+        return result
+
+
+def serialize_kubernetes_model(model: Any) -> Dict[str, Any]:
+    """Serializes a Kubernetes model.
+
+    Args:
+        model: The model to serialize.
+
+    Raises:
+        TypeError: If the model is not a Kubernetes model.
+
+    Returns:
+        The serialized model.
+    """
+    if not is_model_class(model.__class__.__name__):
+        raise TypeError(f"Unable to serialize non-kubernetes model {model}.")
+
+    assert hasattr(model, "attribute_map")
+    attribute_mapping = cast(Dict[str, str], model.attribute_map)
+
+    model_attributes = {
+        serialized_attribute_name: getattr(model, attribute_name)
+        for attribute_name, serialized_attribute_name in attribute_mapping.items()
+    }
+    return _serialize_dict(model_attributes)
+
+
+def _serialize_dict(dict_: Dict[str, Any]) -> Dict[str, Any]:
+    """Serializes a dictionary.
+
+    Args:
+        dict_: Dict to serialize.
+
+    Returns:
+        The serialized dict.
+    """
+    return {key: _serialize(value) for key, value in dict_.items()}
+
+
+def _serialize(value: Any) -> Any:
```

Comment:
This simply seems liek too generic a function to be in `pod_settings.py`. Maybe refactor it out into utils, or rename to `_serialize_kubenertes`